### PR TITLE
Add initial work on Taskcluster training pipeline

### PR DESCRIPTION
**Pull request by bhearsum**
This a large patch, but each commit should be fairly self contained and explained. Overall this PR does the following:
* Adds most (maybe all) of the necessary dependencies for the pipeline -- things like the CUDA toolkit, Marian, et. al.
* Adds kinds for the few stages of the pipeline: dataset fetching, cleaning, and bicleaner/bicleaner-ai. There's going to be some tweaks needed for these still most likely - but in general work in general.
* Adds a `train` action task that allows the `clean` or `bicleaner` stages to be run. (This action task needs a lot of improvement still...but it's a starting point.)
* Adds support for level 3 tasks, largely so that we can start running "real" tasks on pushes, and iterate further in PRs without the risk of polluting the indexes between the two.

I'll also note that this depends on my taskgraph repo at the moment, although there's very little there that we need at this point. (In an earlier iteration I was using some heavily modified fetch task pieces - but that has since been dropped.)

### CODE REVIEW

**Comment by bhearsum** [comment]
This is probably ready for review now. @ahal - you're probably the best RelEng person to have a look seeing as you reviewed the earlier iteration. There's things in here (mostly noted in comments) that I ended up to clean up a bit in the near future, but I can also block merging this to `main` on whatever cleanup you feel is necessary.

@andrenatal - At the very least, I imagine you should sanity check the pipeline scripts that I'm modifying. We can also talk through the other parts as much as you want next time we chat.

File: taskcluster/requirements.in
Comment (ahal) [inline taskcluster/requirements.in:1]: Ah, I guess this is why you disabled hashes :p 

Don't forget to revert this before landing.

**Review by ahal — CHANGES_REQUESTED** [review]
Looks great, nice work! I didn't look too closely at the stuff in `pipeline`.

Don't have anything major, though this is a big enough patch that I'll request changes instead of r+ with nits.

File: .taskcluster.yml
Comment (bhearsum) [reply .taskcluster.yml:79]: No, I think this would apply to PRs created from anywhere come to think of it....(head repo is the one containing the PR branch and base is the repo the PR is being opened against, I think?).

Seems like this is a longstanding oversight...

File: taskcluster/translations_taskgraph/transforms/cache.py
Comment (bhearsum) [reply taskcluster/translations_taskgraph/transforms/cache.py:29]: Yeah, I was wondering about as well. I'm happy to upstream it once it stabilizes.

File: taskcluster/translations_taskgraph/transforms/dataset_substitutions.py
Comment (bhearsum) [reply taskcluster/translations_taskgraph/transforms/dataset_substitutions.py:48]: I wasn't necessarily planning to do that prior to landing this, but I could.

File: taskcluster/requirements.in
Comment (bhearsum) [reply taskcluster/requirements.in:1]: Yeah, that's right :). Will do.

File: .taskcluster.yml
Comment (bhearsum) [reply .taskcluster.yml:79]: Rereading, and I think you're right -- actions on PRs that were creating from a branch on the canonical repo would get level 3 for actions. (head repo would be the canonical one only for those - it would be a different repo for PRs like this)

Regardless, I agree that it's not ideal that actions running against a PR get level 3, when the rest of the jobs are level 1. Do we have a way to address that though? I'm not sure we have any information about PR vs. push when we render this for `task_for = "action"`...

File: taskcluster/ci/dataset/kind.yml
Comment (bhearsum) [reply taskcluster/ci/dataset/kind.yml:45]: Indeed - this was leftover from earlier. I'll remove it.

File: taskcluster/docker/toolchain-build/Dockerfile
Comment (eu9ene) [inline taskcluster/docker/toolchain-build/Dockerfile:26]: @bhearsum @andrenatal I'm curious whether we want to preserve running locally using Snakemake and running on Slurm. If yes, should we add `zstandard` to `envs/base.yml`?

Comment (bhearsum) [reply taskcluster/docker/toolchain-build/Dockerfile:26]: That's an excellent question. I think we certainly need to be able to run both for a period of time. The changes to the pipeline scripts are simple enough (thanks to the command line options being the same) that perhaps we can support both there?

Comment (bhearsum) [reply taskcluster/docker/toolchain-build/Dockerfile:26]: (An alternative would be to run either Taskcluster or the Slurm pipeline off of a separate branch.)

File: .taskcluster.yml
Comment (ahal) [reply .taskcluster.yml:79]: Yeah I don't think there is. Maybe we could derive the level from the Decision task and create a new `actionLevel` context here or something:
https://hg.mozilla.org/ci/ci-configuration/file/tip/build-decision/src/build_decision/util/trigger_action.py#l126

I'm unsure if this is a big deal or a nothing burger.

Comment (bhearsum) [reply .taskcluster.yml:79]: Well...you still need scopes to do anything interesting or sensitive, so I don't think it's a security issue. I think it's mostly a confusion thing, unless I'm missing something.

File: taskcluster/docker/toolchain-build/Dockerfile
Comment (bhearsum) [reply taskcluster/docker/toolchain-build/Dockerfile:26]: Echoing some discussion from Slack -- it sounds like it's useful in the long term to maintain support for gzip to make it easier for others to run our scripts/pipeline. I'm going to update this to continue to use gz as the default compression mechanism, with a way to override it with zstd.

**Comment by bhearsum** [comment]
I think this is ready for anothe review pass -- all of the comments that have been resolved should be addressed (some were merged in with earlier comments, others are in newer commits - sorry for that confusion). I don't think any of the remaining open comments need any action now, or maybe at all. (If any of that is wrong just let me know and I'll take another pass.)

@ahal - I want to specifically call out the change I made to `bicleaner/kind.yml`. I ended up leaving the command in the kind, but I added a new transform that can bring parameters into `command-context` -- which avoids the need for double formatting. Do you still think it's worthwhile pushing more data into transforms?

(I know that GitHub thinks the Taskcluster tests aren't done...but they are if you look at the graph. I think there's some infra issues with GitHub right now.)

**Review by ahal — APPROVED** [review]
Thanks, looks like a good base to build off of!

**Comment by bhearsum** [comment]
@eu9ene - How do you feel about the latest version re: pipeline scripts? It _should_ mean that nothing changes for any Snakemake/Slurm setups.

**Comment by eu9ene** [comment]
> @eu9ene - How do you feel about the latest version re: pipeline scripts? It _should_ mean that nothing changes for any Snakemake/Slurm setups.

lgtm. Abstraction from the compression format is a good idea until it's clear that we can drop compatibility with gzip.

File: pipeline/clean/clean-mono.sh
Comment (bhearsum) [reply pipeline/clean/clean-mono.sh:31]: Yes! Sorry for the oversight.

**Comment by bhearsum** [comment]
@andrenatal and I spoke elsewhere about this PR earlier. He authorized me to go ahead and merge this to unblock the necessary follow-up work (implementing more of the pipeline; experimenting with running the tasks already implemented).

**Pull request merged**
